src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Handlers/Fx2Nl/Fx2NlLanguageServerOperationHandler.cs

```diff
@@ -0,0 +1,102 @@
+    /// </summary>
+    internal sealed class Fx2NlLanguageServerOperationHandler : ILanguageServerOperationHandler
+    {
+        public bool IsRequest => true;
+
+        public string LspMethod => CustomProtocolNames.FX2NL;
+
+        private readonly INLHandlerFactory _nLHandlerFactory;
+
+        private CheckResult _preHandleCheckResult;
+
+        private Fx2NLParameters _fx2NlParameters;
+
+        private CustomFx2NLParams _fx2NlRequestParams;
+
+        private NLHandler _nLHandler;
+
+        public Fx2NlLanguageServerOperationHandler(INLHandlerFactory nLHandlerFactory)
+        {
+            _nLHandlerFactory = nLHandlerFactory;
+        }
+
+        /// <summary>
+        /// Performs pre-handle operations for Fx2Nl.
+        /// </summary>
+        /// <param name="operationContext">Language Server Operation Context.</param>
+        /// <param name="cancellationToken">Cancellation Token.</param>
+        private Task PreHandleFx2NlAsync(LanguageServerOperationContext operationContext, CancellationToken cancellationToken)
+        {
+            _nLHandler = operationContext.GetNLHandler(_fx2NlRequestParams.TextDocument.Uri, _nLHandlerFactory, _fx2NlRequestParams) ?? throw new NullReferenceException("No suitable handler found to handle Fx2Nl");
+            if (!_nLHandler.SupportsFx2NL)
+            {
+                throw new NotSupportedException("Fx2Nl is not supported");
+            }
+
+            return operationContext.ExecuteHostTaskAsync(
+            _fx2NlRequestParams.TextDocument.Uri,
+            (scope) =>
+            {
+                _preHandleCheckResult = scope?.Check(_fx2NlRequestParams.Expression) ?? throw new NullReferenceException("Check result was not found for Fx2NL operation");
+                _fx2NlParameters = GetFx2NlHints(scope, operationContext);
+            }, cancellationToken);
+        }
+
+        /// <summary>
+        /// Gets the Fx2Nl hints.
+        /// </summary>
+        /// <param name="scope"> PowerFx Scope.</param>
+        /// <param name="operationContext">Language Server Operation Context.</param>
+        /// <returns>Fx2Nl hints.</returns>
+        private Fx2NLParameters GetFx2NlHints(IPowerFxScope scope,  LanguageServerOperationContext operationContext)
+        {
+            return scope is IPowerFxScopeFx2NL fx2NLScope ? fx2NLScope.GetFx2NLParameters() : new Fx2NLParameters();
+        }
+
+        /// <summary>
+        /// Performs core Fx2Nl operation.
+        /// </summary>
+        /// <param name="operationContext">Language Server Operation Context.</param>
+        /// <param name="cancellationToken">Cancellation Token.</param>
+        private async Task Fx2NlAsync(LanguageServerOperationContext operationContext, CancellationToken cancellationToken)
+        {
+            var result = await _nLHandler.Fx2NLAsync(_preHandleCheckResult, _fx2NlParameters, cancellationToken).ConfigureAwait(false);

```

Comment:
> _nLHandler.Fx2NLAsync(_preHandleCheckResult, [](http://example.com/codeflow?start=30&length=45)

Subtle bug - usage of checkREsult should be done under the lock. 
[https://msazure.visualstudio.com/DefaultCollection/OneAgile/_git/Power%20Fx%20Internal?path=/src/Microsoft.PowerFx.NaturalLanguage/Helper/LanguageServerExtensions.cs&version=GBmain&line=98&lineEnd=99&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents](https://msazure.visualstudio.com/DefaultCollection/OneAgile/_git/Power%20Fx%20Internal?path=/src/Microsoft.PowerFx.NaturalLanguage/Helper/LanguageServerExtensions.cs&version=GBmain&line=98&lineEnd=99&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents)

So CheckResult shouldn't be the field, just the Fx2NLPayload that we computed from it.

Reply:
We talked offline and i think this is not a blocker and we should do an immediate follow up on this?